src/Interpreters/GinFilter.cpp

```diff
@@ -0,0 +1,224 @@
+
+        store->getPostings()[token] = builder;
+    }
+}
+
+void GinFilter::addRowRangeToGinFilter(UInt32 segmentID, UInt32 rowIDStart, UInt32 rowIDEnd)
+{
+    if (!rowid_range_container.empty())
+    {
+        /// Try to merge the rowID range with the last one in the container
+        if (rowid_range_container.back().segment_id == segmentID &&
+            rowid_range_container.back().range_end+1 == rowIDStart)
+        {
+            rowid_range_container.back().range_end = rowIDEnd;
+            return;
+        }
+    }
+    rowid_range_container.push_back({segmentID, rowIDStart, rowIDEnd});
+}
+
+void GinFilter::clear()
+{
+    rowid_range_container.clear();
+}
+
+#ifndef NDEBUG
+void GinFilter::dump() const
+{
+    printf("filter : '%s', row ID range:\n", getMatchString().c_str());
+    for (const auto & rowid_range: rowid_range_container)
+    {
+        printf("\t\t%d, %d, %d; ", rowid_range.segment_id, rowid_range.range_start, rowid_range.range_end);
+    }
+    printf("\n");
+}
+
+void dumpPostingsCache(const PostingsCache& postings_cache)
+{
+    for (const auto& term_postings : postings_cache)
+    {
+        printf("--term '%s'---------\n", term_postings.first.c_str());
+
+        const SegmentedPostingsListContainer& container = term_postings.second;
+
+        for (const auto& [segment_id, postings_list] : container)
+        {
+            printf("-----segment id = %d ---------\n", segment_id);
+            printf("-----postings-list: ");
+            for (unsigned int it : *postings_list)
+            {
+                printf("%d ", it);
+            }
+            printf("\n");
+        }
+    }
+}
+
+void dumpPostingsCacheForStore(const PostingsCacheForStore& cache_store)
+{
+    printf("----cache---store---: %s\n", cache_store.store->getName().c_str());
+    for (const auto & query_string_postings_cache: cache_store.cache)
+    {
+        printf("----cache_store----filter string:%s---\n", query_string_postings_cache.first.c_str());
+        dumpPostingsCache(*query_string_postings_cache.second);
+    }
+}
+#endif
+
+bool GinFilter::hasEmptyPostingsList(const PostingsCachePtr& postings_cache)
+{
+    if (postings_cache->empty())
+        return true;
+
+    for (const auto& term_postings : *postings_cache)
+    {
+        const SegmentedPostingsListContainer& container = term_postings.second;
+        if (container.empty())
+            return true;
+    }
+    return false;
+}
+
+bool GinFilter::matchInRange(const PostingsCachePtr& postings_cache, UInt32 segment_id, UInt32 range_start, UInt32 range_end)
+{
+    /// Check for each terms
+    GinIndexPostingsList intersection_result;
+    bool intersection_result_init = false;
+
+    for (const auto& term_postings : *postings_cache)
+    {
+        const SegmentedPostingsListContainer& container = term_postings.second;
+        auto container_it{ container.find(segment_id) };
+        if (container_it == container.cend())
+        {
+            return false;
```

Comment:
My understanding that you have segments, you should not prematurely return but rather continue looping through the postings_cache, am I wrong? So if segment_id is not in the first postings_cache, you bail out. I don't see that to be right

Reply:
All segment info are loaded together, so all segment ids for one part are either in cache or not in cache. see GinIndexStoreFactory::get().